README.md
<h1 align="center"> NDIScopes</h1>
<p align="center">
Use several different <a href="https://github.com/MindStudioOfficial/ndiscopes#waveformsscopes">Waveforms/Scopes</a> to analyze a <a href="https://ndi.tv">NDI®</a> Stream. 
</p>

<div align="center">

<a href="https://github.com/MindStudioOfficial/ndiscopes/releases">
<img src="https://img.shields.io/github/downloads-pre/MindStudioOfficial/ndiscopes/total?style=flat-square&label=Downloads"></a>
<a href="https://github.com/MindStudioOfficial/ndiscopes/releases">
<img src="https://img.shields.io/github/downloads-pre/MindStudioOfficial/ndiscopes/latest/total?style=flat-square&label=Downloads@latest"></a>
<!--<a href="">
<img src="https://img.shields.io/tokei/lines/github/MindStudioOfficial/ndiscopes"></a>-->

</div>
<div align="center">

<a href="https://github.com/MindStudioOfficial/ndiscopes/releases">
<img src="https://img.shields.io/github/v/release/MindStudioOfficial/ndiscopes?style=flat-square&include_prereleases&label=Version"></a>


</div>
<div align="center">

<a href="https://github.com/MindStudioOfficial/ndiscopes">
<img src="https://img.shields.io/github/stars/MindStudioOfficial/ndiscopes?style=flat-square"></a>
<a href="https://github.com/MindStudioOfficial/ndiscopes">
<img src="https://img.shields.io/github/watchers/MindStudioOfficial/ndiscopes?style=flat-square"></a>
</div>
<br>
<hr>
<div align="center">
<h3>We've re-branded as "LiveScopes.tv"</h3>
<a href="https://github.com/MindStudioOfficial/livescopestv">New GitHub Page</a>
</div>

<hr>

![Screenshot](blob/sc7.jpg)

| Luminance: ![Luma Waveform](blob/sc_luma.jpg)                      | RGB Overlayed: ![RGB Waveform](blob/sc_rgb.jpg)                      | RGB Parade: ![RGB Parade](blob/sc_rgbParade.jpg)                        |
| ------------------------------------------------------------------ | -------------------------------------------------------------------- | ----------------------------------------------------------------------- |
| **15% RGB Blacklevel:**  ![Black Level](blob/sc_blacklevel.jpg)    | **YCbCr Parade:** ![YUV Parade](blob/sc_yuvparade.jpg)               | **YUV Vectorscope:** ![Vectorscope](blob/sc_vectorscope.jpg)            |
| **With Alpha background:**  ![Alpha Background](blob/sc_alpha.jpg) | **Without Alpha Background:**![Alpha Background](blob/sc_alpha2.jpg) | **Audiometers dBu:** <br><img src="blob/sc_audiometers.jpg" height=200> |

| False Color: ![False Color](blob/sc_falseColor2.jpg) | Splitscreen:![Splitscreen](blob/sc_splitscreen.jpg) |
| ---------------------------------------------------- | --------------------------------------------------- |

## Requirements

The software uses **CUDA** to compute the frames and scopes. A **NVIDIA GPU** is required to run this software. See [GPU Support](https://github.com/MindStudioOfficial/ndiscopes#gpu-support) to see exaxtly what GPUs are supported.

- OS: 
  | Windows | Tested |
  | ------- | ------ |
  | 11      | ✔      |
  | 10      | ✔      |
- CPU: x64, >= 4 Cores
- GPU: See [GPU Support](https://github.com/MindStudioOfficial/ndiscopes#gpu-support)
- RAM: Software uses at max 2GB (rarely)
- NVIDIA Driver: >= 452.39 (522.06 recommended)
- Networking: NDI® works best on a wired >=1GBit/s ethernet link.

## Platform Support

![Windows](https://upload.wikimedia.org/wikipedia/commons/thumb/4/48/Windows_logo_-_2012_%28dark_blue%29.svg/88px-Windows_logo_-_2012_%28dark_blue%29.svg.png)

## Features

### Waveforms/Scopes
- [x] Luminance Waveform
- [x] RGB Waveform
- [x] RGB Parade
- [x] CbCr/UV Vectorscope
- [x] False Color
- [x] YUV Parade
- [x] CIE 1931 Chromaticity
- [x] RGBA Channel Isolation

### Settings
- [x] Scale vectorscope x0.5 to x5
- [x] Toggle waveform scale labels
- [x] Waveform scale as percentage
- [x] Waveform scale as 8Bit value
- [x] Save and load settings on startup
- [x] Enable Audiometer 
- [x] Select Audio Device

### Audio
- [x] Audiometer for every audio channel of the NDI® source in dBu scale
- [x] Audio playback to default audio device (toggleable) 
- [x] Selectable audio device

### Other
- [x] Reference Frame
  - [x] Save/Load
  - [x] Overlay on source as splitscreen
  - [x] Overlay on scopes as splitscreen
  - [x] Show thumbnails in frame browser
- [x] Draw Masks on source frame
- [x] View Timecode
- [x] View Embeded Metadata
### NDI® Codec support
- [x] UYVY (most common, fastest)
- [x] BGRA (stills and alpha)

## Keyboard Shortcuts

- <kbd>A</kbd> Toggle `A`lpha Grid
- <kbd>B</kbd> Toggle Frame `B`rowser
- <kbd>C</kbd> Toggle False `C`olor
- <kbd>D</kbd> Toggle Split `D`irection
- <kbd>F</kbd> `F`lip Split Side
- <kbd>M</kbd> Toggle `M`ask
- <kbd>S</kbd> Select `S`ource
- <kbd>T</kbd> Toggle `T`imecode
- <kbd>X</kbd> Toggle Settings
- <kbd>+</kbd> Capture new Reference Frame
- <kbd>Del</kbd> Disable Reference Frame Overlay

(Feel free to suggest alternative shortcuts as a Feature Request Issue)

## GPU Support
NVIDIA Cards only:
  | CC      | Name                 | Cards                                                               | Compatible | Tested |
  | ------- | -------------------- | ------------------------------------------------------------------- | ---------- | ------ |
  | 9.0     | Hopper               | H100                                                                | ✔          | ?      |
  | 8.9     | Ada Lovelace         | RTX 40 Series, RTX 6000                                             | ✔          | ?      |
  | 8.0-8.7 | Ampere               | RTX 30 Series, RTX A Series                                         | ✔          | ✔      |
  | 7.x     | Volta/Turing         | RTX 20 Series, GTX 16 Series, Quadro RTX Series, TITAN RTX, TITAN V | ✔          | ✔      |
  | 6.x     | Pacal                | GTX 10 Series, Titan X, Quadro P Series                             | ✔          | ✔      |
  | 5.x     | Maxwell              | GTX 750 - GTX 980 Ti                                                | ✔          | ?      |
  | >= 3.5  | Kepler               | GT 640 - GTX 780 Ti, TITAN Z                                        | ✔          | ?      |
  | < 3.5   | Kepler, Fermi, Tesla |                                                                     | ❌          |        |

  

## Download

Download the latest version under [Releases](https://github.com/MindStudioOfficial/ndiscopes/releases).


## License

[End User License Agreement](LICENSE.md)

## Author

developed by [**Marc Bach** _"MindStudio"_](https://github.com/MindStudioOfficial/)

<hr> 

*NDI® is a registered trademark of Vizrt Group.*

CHANGELOG.md
# Changelog

## [v0.8.1-beta] 2023-01-23

### Added
- 🧙🏼‍♂️ Added Installer 
- 📜 Added End-User-License-Agreement `EULA.txt`
- 📃 Added About Text in About Menu

### Changed
- 🚀 Improved rendering performance by optimizing GPU memory management
- 🐞 Fixed: Application not shutting down completely and staying active in background in some cases
- ⌨ Changed keyboard shortcuts to 
  - <kbd>T</kbd> Timecode
  - <kbd>A</kbd> Alpha Grid
- ⏫ Upgraded NDI SDK from 5.5.1 to 5.5.2

## [v0.8.0-beta] 2022-11-06

### Added
- 🎨 Added colorspace differentiation for vectorscope and CIE chromaticity scope
- 🎨 Added rec601 colorspace to CIE 1931 scope
### Changed
- 🚀 Improved window animations for settings, metadata, and colorfilters
-  ↕ Zoomed in the CIE 1931 scope
- 📏 Vectorscope lines now correspond to the primary and secondary color vectors
- 🔳 Mask now works on stills
- 🔳 Mask background now stays visible
- 🚀 Improved mask performance
### Fixed
- 🐞 Fixed #9 metadata table selection not working correctly
- 🐞 Fixed wrong calculation of vectorscope squares

## [v0.7.3-beta] 2022-11-01

### Added
- Implemented request #10: Fractional Aspect Ratio e.g. **16:9**
### Changed
- Fixed #6: startup failure "Error Loading Audio Devices" when VB-Audio Virtual Cable is present.

## [v0.7.2-beta] 2022-10-20

### Added
- Added Support for GPU Compute Capability 8.9 and 9.0 "Ada Lovelace" (RTX 40 Series and RTX 6000) and "Hopper" (H100)
  - Please report in the discussion of this release if this software works on the GPUs I have [not tested yet](https://github.com/MindStudioOfficial/ndiscopes#gpu-support)
- Added Metadata Table with more data about a NDI Source based on #5. Currently includes:
  - Source Name
  - Source IP Address
  - Resolution
  - Framerate
  - Aspect Ratio
  - FourCC
- Improved performance due to updated rendering and removal of unnecessary memory zeroing
- Added window title bar on loading screen to close the app in case of an error message.

### Changed
- Fixed: GPU Version not checked correctly 
- Fixed: Texture exception on startup

## [v0.7.1-beta] 2022-09-12

### Changed
- Implement better way to wait for NDI/Audio/Renderer to shut down resulting in faster closing time
- Added Links to Bug Reports, Feature Requests, and Discussions to about/help-menu
- Add GPU Compute Capability (GPU Version) to about/help-menu


## [v0.7.0-beta] 2022-09-06

### Added
- ✨ CIE 1931 Chromaticity Scope
- 🎨 Color filter matrix with presets for color/alpha channel isolation and more
- 📁 Open file location button in framebrowser
- ❓ Help/About button with some useful links

### Fixed
- 💻 Fixed scrolling with laptop trackpad guestures not working

### Changed
- ⏫ Upgraded to NDI 5.5
- Adjusted layout of the framebrowser
- Saved reference frame thumbnails are now cached correctly to avoid re-rendering
- Improved error handling on startup

## [v0.6.2-beta] 2022-08-29

### Added
- View timecode (in milliseconds not frames)
- View embedded XML-metadata


## [v0.6.1-beta] 2022-08-02

### Added
- Added option in config to enable accurate rendering of scopes:
- Reduces noise at the cost of performance
  
### Changed
- Fixed: black lines caused by rounding errors in scopes
- Fixed: #7 stuck on loading settings



## [v0.6.0-beta] 2022-07-22

### Added
- ✨New Waveform: YUV Parade
- ✨New Waveform: 15% RGB Black Level
- Modular waveform rendering increases performance because only selected scopes need to be rendered
- Waveform type is now switchable in a dropdown menu above each scope
- Waveform overlay now customizable for each type in a seperate menu ●●●
- Visual improvements and hover/focus feedback to buttons and menus
- Menus can now be traversed in the correct order using <kbd>Tab</kbd> and the arrow keys
- ⌨ Added keyboard shortcuts (see Readme) for several actions
- 🔊 Audio output device is now selectable via the settings menu

### Changed
- Fixed: Framebrowser not listening to File changes on first start of the app
- Fixed: Reference Frame not visible when no input present
- Fixed: NDI Reconnect when saving reference frame (causing lag)
- Fixed: DRPC not updating when

## [v0.5.1-beta] 2022-07-08

### Fixed
- Fix scope layout initializing to all luminance scopes when updating from `v0.4.1`

If you have started the version `v0.5.0` you have to change the layout back manually in the `config.json` located at:
`C:\Users\<username>\AppData\Roaming\com.mindstudio\ndiscopes`

Here change the value of `scopeLayout` from `[0,0,0]` to `[0,1,2]`.

## [v0.5.0-beta] 2022-07-08

### Changed
- Changed Image rendering from using the slow CustomPaint to the much faster Texture workflow:

This Improves Performance by up to 400% (4 times the amount of frames processed per second)

| Resolution  | CustomPaint | Texture   | Average<br>Improvement |
| ----------- | ----------- | --------- | ----------- |
| 1280 x 720  | 8 - 20 ms<br>50 - 125 fps | 3 - 6ms<br>170 - 330 fps  | 300%        |
| 1920 x 1080 | 15 - 50ms<br>20 - 70 fps   | 7 - 10ms<br>100 - 140 fps  | 350%        |
| 2560 x 1440 | 40 - 70ms<br>15 - 25 fps   | 12 - 20ms<br>50 - 85 fps | 340%        |
| 3840 x 2160 | 80 - >200ms<br>5 - 12 fps | 20 - 50ms<br>20 - 50 fps | 400%        |

Thanks to @alexmercerind and the people from the FlutterDev Discord for helping with the implementation.

For more detail and how these numbers were aquired look at this pullrequest #6.

- Other performance improvements in the user interface
- More responsive Sourceselector
- Option to stop receiving frames in Sourceselector
- Vectorscope is now colorized
- Audiometers are now colorful 
- Added performance statistics: 
  - image resolution
  - framerates (actual vs. metadata)
  - renderdelay

### Fixed
- Luminance Scope: Values from 16 to 235 are now mapped to the whole range of 0 - 255
- Audiometers were not visible if no audio data is received
- Vectorscope: Colored Squares were at the wrong saturation levels. (Now 100% and 75%)

### Other
- Added debugging messages for developers

lib/models/colors.dart
import 'package:flutter/material.dart';

Color cDominant = const Color.fromRGBO(14, 14, 14, 1);
Color cPrimary = const Color.fromRGBO(24, 24, 24, 1);
Color cAccent = const Color.fromARGB(255, 15, 29, 43);

Color cHighlight = const Color.fromARGB(255, 25, 120, 215);
Color cFocused = const Color.fromARGB(255, 0, 75, 149);

Color cWindowTitleBar = cDominant;
Color cAppBackground = cDominant;

Color cDialogBackground = cDominant.withOpacity(.8);
Color cSourceCard = cPrimary;
//Color cScopeBorder = cAccent;
Color cScopeTitleBackground = const Color.fromRGBO(22, 22, 22, 1);
Color cScopeTitleBackgroundHover = const Color.fromRGBO(16, 16, 16, 1);
Color cDecorationBorder = cPrimary;
Color cFrameBrowserHeader = cPrimary;
//Color cDirBackground = const Color.fromRGBO(22, 22, 22, 1);
//Color cDirHover = cPrimary;

Map<Color, String> falseColors = {
  const Color.fromRGBO(80, 39, 81, 1): "<2",
  const Color.fromRGBO(9, 101, 150, 1): "2 - 8",
  const Color.fromRGBO(18, 133, 152, 1): "8 - 15",
  const Color.fromRGBO(66, 163, 169, 1): "15 - 24",
  const Color.fromRGBO(133, 133, 133, 1): "24 - 43",
  const Color.fromRGBO(98, 185, 70, 1): "43 - 47",
  const Color.fromRGBO(159, 159, 159, 1): "47 - 54",
  const Color.fromRGBO(236, 181, 188, 1): "54 - 58",
  const Color.fromRGBO(209, 209, 209, 1): "58 - 77",
  const Color.fromRGBO(240, 231, 140, 1): "77 - 84",
  const Color.fromRGBO(255, 255, 1, 1): "84 - 93",
  const Color.fromRGBO(255, 139, 0, 1): "93 - 100",
  const Color.fromRGBO(255, 0, 0, 1): ">=100",
};
